### PR TITLE
Advanced AutoScore Code

### PR DESCRIPTION
**Pull request by Mason-Beard**
If we are at the right RPM: Shoot for allotted time ronan gives in the parameter. it will also move the amp motor slightly so it is pushed all the way back before shooting
If we are not after a certain amount of time that ronan sets (aka withTimeout in the parameter), then 1 of 2 things happen. Either:
a) shooter is not running: we eject the note through the intake
b) shooter is running below the threshold(this means that it is in between 0-setpoint rpm): In this case, we will still shoot it in case we still make it, but print an error message saying that we aren't making the RPM requirement

### CODE REVIEW

File: src/main/java/frc/robot/RobotContainer.java [hunk 1]
```diff
@@ -253,7 +254,12 @@ public void initAutoChooser() {
             AMP.STATE.INTAKING.get(),
             RPM_SETPOINT.SPEAKER.get(),
             INTAKE.STATE.FRONT_ROLLER_INTAKING.get(),
-            INTAKE.STATE.BACK_ROLLER_INTAKING.get()));
+            INTAKE.STATE.BACK_ROLLER_INTAKING.get(),
+            WAIT.SHOOTING.get(),
+            5));
+
+    m_autoChooser.addOption(
+        "ScoreSpeakerTesting", new ScoreSpeaker(m_shooter, m_ampShooter, m_intake));
```
Comment (SherazF) [inline src/main/java/frc/robot/RobotContainer.java:262]: ScoreSpeakerTest?

File: src/main/java/frc/robot/commands/autos/FourPieceNear.java [hunk 2]
```diff
@@ -56,7 +57,9 @@ public FourPieceNear(
             AMP.STATE.INTAKING.get(),
             RPM_SETPOINT.SPEAKER.get(),
             INTAKE.STATE.FRONT_ROLLER_INTAKING.get(),
-            INTAKE.STATE.BACK_ROLLER_INTAKING.get());
+            INTAKE.STATE.BACK_ROLLER_INTAKING.get(),
+            WAIT.SHOOTING.get(),
+            5);
```
Comment (SherazF) [inline src/main/java/frc/robot/commands/autos/FourPieceNear.java:62]: What is this 5 used for?

File: src/main/java/frc/robot/commands/autos/ScoreSpeaker.java [hunk 3]
```diff
@@ -1,22 +1,37 @@
...
-        new AutoSetRPMSetpoint(shooter, RPM_SETPOINT.SPEAKER.get())
-            .withTimeout(WAIT.WAIT_FOR_FLYWHEEL_SETPOINT.get()),
-        new AutoSetRPMSetpoint(shooter, RPM_SETPOINT.NONE.get())
-        // TODO: Replace with Amp Rollers
-        // new AutoRunUptake(uptake, UPTAKE_STATE.DOWNTAKE)
-        );
+        // new AutoRunAmpTake(
+        //         intake,
+        //         ampShooter,
+        //         INTAKE.STATE.FRONT_ROLLER_INTAKING.get(),
```
Comment (SherazF) [inline src/main/java/frc/robot/commands/autos/ScoreSpeaker.java:19]: Can we delete this commented code if it's not needed?

File: src/main/java/frc/robot/commands/shooter/AutoScore.java [hunk 4]
```diff
@@ -15,6 +16,17 @@ public class AutoScore extends Command {
   private final double m_FrontIntakePercentOutput;
   private final double m_BackIntakeAmpPercentOutput;
   private final double allowableError = 2000; // in RPM
+  private final double m_timeToShoot;
+  private final double m_withTimeout;
+  private final double reverseTimerThreshold = 0.25;
+
+  private final double minRPMThreshold = 300;
+
+  private final boolean funny = true;
```
Comment (SherazF) [inline src/main/java/frc/robot/commands/shooter/AutoScore.java:25]: Why is this in the pull request? LOL

File: src/main/java/frc/robot/commands/shooter/AutoScore.java [hunk 5]
```diff
@@ -23,31 +35,84 @@ public AutoScore(
       double AmpPercentOutput,
       double RPMOutput,
       double FrontIntakeAmpPercentOutput,
-      Double BackIntakeAmpPercentOutput) {
+      Double BackIntakeAmpPercentOutput,
+      double TimeToShoot,
+      double withTimeout) {
```
Comment (SherazF) [inline src/main/java/frc/robot/commands/shooter/AutoScore.java:40]: Try to keep caps consistent with camelcase (e.g. timeToShoot)

File: src/main/java/frc/robot/commands/shooter/AutoScore.java [hunk 6]
```diff
@@ -23,31 +35,84 @@ public AutoScore(
...
     addRequirements(m_shooter);
   }
 
   // Called when the command is initially scheduled.
   @Override
-  public void initialize() {}
+  public void initialize() {
+    m_timer.stop();
+    m_timer.reset();
+    m_reversetimer.stop();
```
Comment (SherazF) [inline src/main/java/frc/robot/commands/shooter/AutoScore.java:59]: I'm not familiar with timer logic, but I'll trust that this is correct on initialize?

File: src/main/java/frc/robot/commands/shooter/AutoScore.java [hunk 7]
```diff
@@ -23,31 +35,84 @@ public AutoScore(
...
+    m_reversetimer.reset();
+    m_shoottimer.stop();
+    m_shoottimer.reset();
+  }
 
   @Override
   public void execute() {
     m_shooter.setRPMOutput(m_RPMOutput);
+    m_timer.start();
```
Comment (SherazF) [inline src/main/java/frc/robot/commands/shooter/AutoScore.java:68]: You don't start the shoot or reverse timers here?

File: src/main/java/frc/robot/commands/shooter/AutoScore.java [hunk 8]
```diff
@@ -23,31 +35,84 @@ public AutoScore(
...
+          || (m_shooter.getRpmFollower() < minRPMThreshold)) {
+
+        m_ampShooter.setPercentOutput(-m_AmpPercentOutput);
+        m_intake.setSpeed(-m_FrontIntakePercentOutput, -m_BackIntakeAmpPercentOutput);
+        m_shoottimer.start();
+
+        System.err.println("Flywheel Did not Move");
+        SmartDashboard.putString("isFlyWheelMoving?", "Flywheel Did not Move");
+
+        // Abort Shot bc flywheel is broken
```
Comment (SherazF) [inline src/main/java/frc/robot/commands/shooter/AutoScore.java:110]: What code is this comment tied to?

File: src/main/java/frc/robot/constants/SHOOTER.java [hunk 9]
```diff
@@ -33,7 +33,8 @@ public final class SHOOTER {
 
   public enum WAIT {
     WAIT_FOR_FLYWHEEL_SETPOINT(3),
-    WAIT_FOR_AMP_SCORE(0.8);
+    WAIT_FOR_AMP_SCORE(0.8),
+    SHOOTING(0.75);
```
Comment (SherazF) [inline src/main/java/frc/robot/constants/SHOOTER.java:37]: Is this from testing?

**Review by SherazF — CHANGES_REQUESTED** [review]
Please address the comments I left, and you should be good to merge

File: src/main/java/frc/robot/commands/autos/FourPieceNear.java (hunk 2 above)
Comment (Mason-Beard) [reply src/main/java/frc/robot/commands/autos/FourPieceNear.java]: Withtimeout for the command

**Pull request merged**